### PR TITLE
Make WindowedValueCoder an Interface

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Implement it in both FullWindowedValueCoder and
ValueOnlyWindowedValueCoder

Make ValueOnlyWindowedValueCoder a CustomCoder

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java [hunk 1]
```diff
@@ -716,25 +714,24 @@ public CloudObject initializeCloudObject() {
   /**
    * Coder for {@code WindowedValue}.
    *
-   * <p>A {@code ValueOnlyWindowedValueCoder} only encodes and decodes the value. It drops
-   * timestamp and windows for encoding, and uses defaults timestamp, and windows for decoding.
+   * <p>A {@code ValueOnlyWindowedValueCoder} only encodes and decodes the value. It drops timestamp
+   * and windows for encoding, and uses defaults timestamp, and windows for decoding.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java]: Fix some minor grammatical issues:
It drops timestamp and windows when encoding, and uses a default timestamp and window when decoding.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java [hunk 2]
```diff
@@ -716,25 +714,24 @@ public CloudObject initializeCloudObject() {
...
-        @JsonProperty(PropertyNames.COMPONENT_ENCODINGS)
-        List<Coder<?>> components) {
-      checkArgument(components.size() == 1, "Expecting 1 component, got " + components.size());
-      return of(components.get(0));
-    }
+    private final Coder<T> valueCoder;
 
     ValueOnlyWindowedValueCoder(Coder<T> valueCoder) {
-      super(valueCoder);
+      this.valueCoder = valueCoder;
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java]: checkNotNull(valueCoder)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java]: Done.

**Comment by lukecwik** [comment]
LGTM
I will merge this.

**Comment by dhalperi** [comment]
Context for this work? (No JIRA).

This broke DataflowRunner and may interfere with Hackathon if not fixed before then. Should we revert?

**Comment by tgroh** [comment]
Reverting is fine. This is part of the drive to clean up some of the coder Hierarchy, but the revert should be compatible with the new approach I'm taking.

**Comment by dhalperi** [comment]
I'll send a simpler fixup PR, perhaps.

**Pull request closed without merging**